src/cata_tiles.h

```diff
@@ -25,9 +25,37 @@ class JsonObject;
 struct visibility_variables;
 
 /** Structures */
+struct sprite_variation {
+    int weight;
+    std::vector<int> sprite_ids;
+
+    sprite_variation()
+    {
+        weight = 1;
+        sprite_ids.clear();
+    }
+
+    sprite_variation(int id)
+    {
+        weight = 1;
+        sprite_ids.clear();
+        sprite_ids.push_back(id);
+    }
+
+    sprite_variation(int w, int id)
+    {
+        weight = w;
+        sprite_ids.clear();
+        sprite_ids.push_back(id);
+    }
+};
+
 struct tile_type
 {
-    std::vector<int> fg, bg;
+    // first entry of each of these vectors is not a real sprite_variation
+    // but a bookkeeping list instead, used for reducing the task of selecting
+    // a random variation from O(N) to O(1)
```

Comment:
> adding something like  vector<int> fg_pick, bg_pick

Yes, logically separated data belongs into separate members.

It looks like `fg[0]` doesn't actually store tile-ids (as all the other elements of `fg` do), but indexes into the sprites vectors or so, is that about right? In that case, it should certainly be a separate member, as it is a completely different type of data. It's just a coincidence that you can use the same C++ type (vector) for it.

This allows you to get rid of the special case for `weight == -1` (for example in `load_tilejson_from_file`). It's much easier to read `curr_tile.fg_bookkeeping.push_back(s)` instead of `curr_tile.fg[0].push_back(s)`. ("Huh, what's so special about the first element here. Why are some of the loops in `load_tilejson_from_file` going over *all* elements, it doesn't seem special there.")

----

I don't get the last sentence. You *did* build the  sprite_variation class from scratch.

You should be able to extend the existing `weighted_list` class for your needs. You probably need to add the bookkeeping vector, filled either by overriding the `add` function or by a dedicated finalization function. The `get` function can be overridden similar to use the bookkeeping data.